test/e2e-api/kafka.go

```diff
@@ -22,201 +22,254 @@ import (
+		return
+	}
 
-		ns := r.URL.Query().Get("n")
-		if ns == "" {
-			ns = "-1"
-		}
-		n, err := strconv.Atoi(ns)
-		if err != nil {
-			http.Error(w, err.Error(), http.StatusBadRequest)
-			return
-		}
+	message := &sarama.ProducerMessage{
+		Topic:     topic,
+		Value:     sarama.ByteEncoder(buf),
+		Key:       sarama.ByteEncoder(key),
+		Partition: int32(partition),
+	}
+	if _, _, err := kh.producer.SendMessage(message); err != nil {
+		_, _ = fmt.Fprintf(w, "ERROR: %v\n", err)
+	}
+}
 
-		w.Header().Set("Content-Type", "application/octet-stream")
-		w.WriteHeader(200)
+func (kh *KafkaController) PumpTopicHandler(w http.ResponseWriter, r *http.Request) {
+	topic := r.URL.Query().Get("topic")
+	mf := newMessageFactory(r.URL.Query())
+	duration, err := time.ParseDuration(r.URL.Query().Get("sleep"))
+	if err != nil {
+		log.Println(err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
 
-		config := sarama.NewConfig()
-		config.Producer.Return.Successes = true
+	ns := r.URL.Query().Get("n")
+	if ns == "" {
+		ns = "-1"
+	}
+	n, err := strconv.Atoi(ns)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
 
-		syncProducer, err := sarama.NewSyncProducer(brokers, config)
+	w.Header().Set("Content-Type", "application/octet-stream")
+	w.WriteHeader(200)
 
-		if err != nil {
-			log.Println(err)
-			http.Error(w, err.Error(), http.StatusInternalServerError)
+	start := time.Now()
+	_, _ = fmt.Fprintf(w, "sending %d messages of size %d to %q\n", n, mf.size, topic)
+
+	for i := 0; i < n || n < 0; i++ {
+		select {
+		case <-r.Context().Done():
 			return
-		}
-		defer syncProducer.Close()
-
-		start := time.Now()
-		_, _ = fmt.Fprintf(w, "sending %d messages of size %d to %q\n", n, mf.size, topic)
-
-		for i := 0; i < n || n < 0; i++ {
-			select {
-			case <-r.Context().Done():
-				return
-			default:
-				message := &sarama.ProducerMessage{
-					Topic: topic,
-					Value: sarama.ByteEncoder(mf.newMessage(i)),
-					Key:   sarama.ByteEncoder([]byte(strconv.Itoa(i))),
-				}
-				_, _, err := syncProducer.SendMessage(message)
-				if err != nil {
-					_, _ = fmt.Fprintf(w, "ERROR: %v\n", err)
-				}
-				time.Sleep(duration)
+		default:
+			message := &sarama.ProducerMessage{
+				Topic:     topic,
+				Value:     sarama.ByteEncoder(mf.newMessage(i)),
+				Key:       sarama.ByteEncoder(strconv.Itoa(i)),
+				Partition: int32(0),
+			}
+			_, _, err := kh.producer.SendMessage(message)
+			if err != nil {
+				_, _ = fmt.Fprintf(w, "ERROR: %v\n", err)
 			}
+			time.Sleep(duration)
 		}
-		_, _ = fmt.Fprintf(w, "sent %d messages of size %d at %.0f TPS to %q\n", n, mf.size, float64(n)/time.Since(start).Seconds(), topic)
-	})
+	}
+	_, _ = fmt.Fprintf(w, "sent %d messages of size %d at %.0f TPS to %q\n", n, mf.size, float64(n)/time.Since(start).Seconds(), topic)
+}
+
+func (kh *KafkaController) Close() {
+	_ = kh.producer.Close()
+	_ = kh.consumer.Close()
+	_ = kh.adminClient.Close()
```

Comment:
Does the order of close statements matter?

Reply:
No